rollouts/controllers/rollout_controller_test.go

```diff
@@ -0,0 +1,373 @@
+				`Eventually()` will repeatedly run the function provided as an argument every interval seconds until
+				(a) the function’s output matches what’s expected in the subsequent `Should()` call, or
+				(b) the number of attempts * interval period exceed the provided timeout value.
+				In the examples below, timeout and interval are Go Duration values of our choosing.
+			*/
+
+			rolloutLookupKey := types.NamespacedName{Name: RolloutName, Namespace: RolloutNamespace}
+			createdRollout := &gitopsv1alpha1.Rollout{}
+
+			// We'll need to retry getting this newly created Rollout, given that creation may not immediately happen.
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, rolloutLookupKey, createdRollout)
+				return err == nil
+			}, timeout, interval).Should(BeTrue())
+			Expect(createdRollout.Spec.Description).Should(Equal("Test Rollout"))
+
+			remoteSyncKey := types.NamespacedName{Name: "github-589324850-namespaces-e2e-sjc-0", Namespace: RolloutNamespace}
+			remoteSync := &gitopsv1alpha1.RemoteSync{}
+
+			// We should eventually have a remotesync object created corresponding to a target cluster
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, remoteSyncKey, remoteSync)
+				return err == nil
+			}, timeout, interval).Should(BeTrue())
+			Expect(remoteSync.Spec.Template.Spec.Git.Repo).To(Equal("https://github.com/droot/store.git"))
+			Expect(remoteSync.Spec.Template.Spec.Git.Revision).To(Equal("v3"))
+
+			// We should eventually have the rollout completed
+			Eventually(func() bool {
+				_ = k8sClient.Get(ctx, rolloutLookupKey, createdRollout)
+				return createdRollout.Status.Overall == "Completed"
+			}, 1*time.Minute, interval).Should(BeTrue())
+
+			Expect(createdRollout.Status.ClusterStatuses).To(HaveLen(1))
+			Expect(createdRollout.Status.ClusterStatuses).To(ContainElement(gitopsv1alpha1.ClusterStatus{
+				Name: "e2e-sjc-0",
+				PackageStatus: gitopsv1alpha1.PackageStatus{
+					PackageID:  "github-589324850-namespaces-e2e-sjc-0",
+					Status:     "Synced",
+					SyncStatus: "Synced",
+				},
+			}))
+			forground := metav1.DeletePropagationForeground
+			Expect(k8sClient.Delete(context.TODO(), createdRollout, &client.DeleteOptions{PropagationPolicy: &forground})).To(Succeed())
+			// We should wait for the rollout to be deleted
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, rolloutLookupKey, createdRollout)
+				return client.IgnoreNotFound(err) == nil
+			}, 1*time.Minute, interval).Should(BeTrue())
+		})
+	})
+
+	Context("A progressive Rollout", func() {
+		// setup target clusters
+		BeforeEach(func() {
+			var err error
+			targets = []e2eclusters.Config{
+				{
+					Prefix: "e2e-sjcc-",
+					Count:  1,
+					Labels: map[string]string{
+						"city": "sjcc",
+					},
+				},
+				{
+					Prefix: "e2e-sfoo-",
+					Count:  1,
+					Labels: map[string]string{
+						"city": "sfoo",
```

Comment:
Why `sjcc` and `sfoo` instead of `sjc` and `sfo`? Is there some conflict with the clusters setup in the previous test? 

Since this is otherwise the same function passed to `BeforeEach` as the previous test, I was wondering if it would make sense to define this func elsewhere and reuse it across tests.

Reply:
We want these two tests to be independent of each other so that we can run them in parallel (though running in parallel requires running with `ginkgo` CLI, but that's a different story). Since we share the APIServer/Etcd between the two tests, keeping a separate `target` clusters is required to keep these tests independent.

Also, each tests may require different target cluster environment (for ex. progressive may have more advance test setup here), so keeping them separate gives that flexibility. So duplication here is the feature not a bug :)